### ISSUE TITLE
Literal not inferred as TypedDict in generator expression

### ISSUE THREAD

**Issue by goodmami** [microsoft/pyright#3948]
**Describe the bug**

Literal dicts produced in a generator expression are not inferred as compatible with declared TypedDicts while they are inferred properly in list comprehensions.

**To Reproduce**

```python
from typing import TypedDict

class X(TypedDict):
    x: str

xs: list[X] = []
xs.extend({"x": c} for c in 'abc')  # "dict[str, str]" is incompatible with "X"
xs.extend([{"x": c} for c in 'abc'])  # no problem in list comprehension
```

**Expected behavior**

I expect the generator expression to behave the same as the list comprehension. In the above MWE, `xs.extend()` expects an argument with type `Iterable[X]`, and both the generator expression and the list comprehension provide that.

**VS Code extension or command-line**

pyright version 1.1.271 on the command line:

```console
$ python -m pyright test.py
No configuration file found.
No pyproject.toml file found.
stubPath /home/goodmami/tst/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
pyright 1.1.271
/home/goodmami/tst/test.py
  /home/goodmami/tst/test.py:7:11 - error: Argument of type "Generator[dict[str, str], None, None]" cannot be assigned to parameter "__iterable" of type "Iterable[X]" in function "extend"
    TypeVar "_T_co@Iterable" is covariant
      "dict[str, str]" is incompatible with "X" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
Completed in 0.526sec
```

**Additional context**

Seems related to #2967.

**Comment by erictraut** [comment]
Thanks for the enhancement request. I've added support for bidirectional type inference for comprehensions, so the above code type checks without error.

This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.272. It will also be included in a future release of pylance.
